shared/src/main/scala/mlscript/NormalForms.scala

```diff
@@ -122,30 +140,35 @@ class NormalForms extends TyperDatatypes { self: Typer =>
       case (RhsBases(ps, bf), p: ClassTag) =>
         S(RhsBases(if (ps.contains(p)) ps else p :: ps , bf))
       case (RhsBases(ps, N), that: MiscBaseType) => S(RhsBases(ps, S(L(that))))
-      case (RhsBases(ps, S(L(TupleType(fs1)))), TupleType(fs2)) =>
-        if (fs1.size =/= fs2.size) N
+      case (RhsBases(ps, S(L(t1@TupleType(fs1)))), t2@TupleType(fs2)) =>
+        if (fs1.size =/= fs2.size) 
+          RhsBases(ps, S(L(t1.toArray))) | t2.toArray // upcast tuple to array
         else S(RhsBases(ps, S(L(TupleType(fs1.lazyZip(fs2).map {
           case ((S(n1), ty1), (S(n2), ty2)) => (if (n1 === n2) S(n1) else N, ty1 | ty2)
           case ((n1o, ty1), (n2o, ty2)) => (n1o orElse n2o, ty1 | ty2)
         })(noProv)))))
+      case (RhsBases(ps, S(L(ArrayType(in)))), TupleType(_)) => N
```

Comment:
Shouldn't this upcast to the common array type? It seems to break the symmetry with other cases.

Reply:
I'm not quite sure about this part now... @_@

Reply:
This case computes `(tag1 | tag2 | tag3 | ... | Array[in]) | (T1, T2, ...)`, which is the same as `tag1 | tag2 | tag3 | ... | (Array[in] | (T1, T2, ...))` and `(Array[in] | (T1, T2, ...)` should reduce to `Array[in | T1 | T2 | ...]`, no?

```suggestion
      case (RhsBases(ps, S(L(ArrayType(in)))), tup @ TupleType(_)) => this | tup.toArray
```